substrate/frame/utility/src/lib.rs

```
@@ -454,6 +464,102 @@ pub mod pallet {
+		///   or both.
+		///
+		/// ## Weight
+		/// The weight of this call is calculated as the sum of:
+		/// `WeightInfo::if_else()` + `main` |
+		/// `WeightInfo::if_else()` + `main` + `fallback`.
+		#[pallet::call_index(6)]
+		#[pallet::weight({
+			let main = main.get_dispatch_info();
+			let fallback = fallback.get_dispatch_info();
+			(
+				T::WeightInfo::if_else()
+					.saturating_add(main.call_weight)
+					.saturating_add(fallback.call_weight),
+				if main.class == Operational && fallback.class == Operational { Operational } else { Normal },
+			)
+		})]
+		pub fn if_else(
+			origin: OriginFor<T>,
+			main: Box<<T as Config>::RuntimeCall>,
+			fallback: Box<<T as Config>::RuntimeCall>,
+		) -> DispatchResultWithPostInfo {
+			// Do not allow the `None` origin.
+			if ensure_none(origin.clone()).is_ok() {
+				return Err(BadOrigin.into());
+			}
+
+			let is_root = ensure_root(origin.clone()).is_ok();
+
+			// Track the weights
+			let mut weight = T::WeightInfo::if_else();
+
+			let info = main.get_dispatch_info();
+
+			// Execute the main call first
+			let main_result = if is_root {
+				main.dispatch_bypass_filter(origin.clone())
+			} else {
+				main.dispatch(origin.clone())
+			};
+
+			// Add weight of the main call
+			weight = weight.saturating_add(extract_actual_weight(&main_result, &info));
+
+			if let Err(main_error) = main_result {
+				// If the main call failed, execute the fallback call
+				let fallback_info = fallback.get_dispatch_info();
+
+				let fallback_result = if is_root {
+					fallback.dispatch_bypass_filter(origin.clone())
+				} else {
+					fallback.dispatch(origin)
+				};
+
+				// Add weight of the fallback call
+				weight =
+					weight.saturating_add(extract_actual_weight(&fallback_result, &fallback_info));
+
+				if let Err(_fallback_error) = fallback_result {
+					// Both calls have faild.
+					return Err(sp_runtime::DispatchErrorWithPostInfo {
+						error: Error::<T>::IfElseBothFailure.into(),
```

not 100% sure about this as it mask the underlying error and makes debugging harder

---

We can alternatively return the fallback error here

```rust
return Err(sp_runtime::DispatchErrorWithPostInfo {
        error: _fallback_error.error,
	post_info: Some(weight).into(),
})
```

---

yeah that will be better than not having anything